### PR TITLE
[modules][expo-document-picker] Support picking multiple documents

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
I've seen this feature being asked for a couple of times on discord or Twitter. 

# How
Added support for picking multiple documents on both platforms 

# Test Plan
Tested the new behaviour in the bare expo app. 

Screenshot_20221207-123006

### CODE REVIEW

**Review by lukmccall — COMMENTED** [review]
The native code looks good, but there is still some work to do.
We want to ensure backward compatibility and make sure that the responses are unified across platforms. 

Can you tell me if we can support multiple documents picking on iOS? If yes, I think we should add it there too. 

Also, I would change the result type. We had a similar problem some time ago with the `ImagePicker`. It was solved by introducing a type that is backward compatible and contains new data. The new type looks like this:
```
{
  oldKey1: ...,
  oldKey2: ...,
  assets: [
    { <new data> }, { ... }, { ... }
   ]
}
```
Here is an example: https://docs.expo.dev/versions/latest/sdk/imagepicker/#imagepickerresult.
The code which is converting those types can be found here: https://github.com/expo/expo/blob/3b094865cca3d0d584861caf1e9659e822865934/packages/expo-image-picker/src/ImagePicker.ts#L73

**Comment by alanjhughes** [comment]
Yes, we can support it on iOS too. Maintaining backward compatibility was my main question. I will follow how it was done in ImagePicker. I'll start the iOS implementation after I fix the types. Had planned on doing that in a separate PR. Should I do them together?

**Comment by lukmccall** [comment]
> Had planned on doing that in a separate PR. Should I do them together?

I think it will make more sense to do them together. We can't merge them separately anyway 🤷

**Comment by alanjhughes** [comment]
That's fine 👍

**Comment by alanjhughes** [comment]
There was a PR to migrate the iOS side of this module to the new API so I waited before doing anything with it. I don't think that has gone any further yet

**Comment by alanjhughes** [comment]
The iOS migration is done bar I need a compiler directive so it runs on iOS 13. Also need some advice on how to deal with the return types. I've looked at how image-picker has done it but it's different here.

File: packages/expo-document-picker/ios/DocumentPickerResponse.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import ExpoModulesCore
+
+internal struct DocumentPickerResponse: Record {
+  @Field var assets: [DocumentInfo]? = nil
+  @Field var canceled: Bool = false
+}
+
+internal struct DocumentInfo: Record {
+  @Field var uri: String = ""
+  @Field var name: String? = nil
```
[the remark this answers is not in the document]
Comment (alanjhughes) [reply packages/expo-document-picker/ios/DocumentPickerResponse.swift:10]: Think I have to do this to initialize a record?

Comment (tsapeta) [reply packages/expo-document-picker/ios/DocumentPickerResponse.swift:10]: Yes, unfortunately it's the downside of records 😞

File: packages/expo-document-picker/src/types.ts [hunk 2]
```diff
@@ -18,56 +18,84 @@ export type DocumentPickerOptions = {
...
+  /**
+   * Document MIME type.
+   */
+  mimeType?: string;
+  /**
+   * Timestamp of last document modification.
+   */
+  lastModified?: number;
+  file?: File;
+  output?: FileList | null;
```
Comment (tsapeta) [inline packages/expo-document-picker/src/types.ts:48]: Could you document them as well?

**Review by tsapeta — APPROVED** [review]

File: packages/expo-document-picker/src/types.ts (hunk 2 above)
Comment (alanjhughes) [reply packages/expo-document-picker/src/types.ts:48]: Will do 👍

**Review by lukmccall — APPROVED** [review]

**Comment by SupriyaPKalghatgi** [comment]
Is this released?

**Pull request merged**